content/news/2023-10-21-bevy-0.12/index.md

```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
+
+Avoiding rebinding is both a big performance benefit for CPU-driven rendering, including WebGL2, and is necessary to enable GPU-driven rendering.
+
+### What are CPU- and GPU-driven rendering?
+
+CPU-driven rendering is where draw commands are created on the CPU, in Bevy this means in Rust code, more specifically in render graph nodes.
+
+In GPU-driven rendering, the draw commands are encoded on the GPU by compute shaders. This leverages GPU parallelism, and unlocks more advanced culling optimisations that are infeasible to do on the CPU, among many other methods that bring large performance benefits.
+
+### Reorder Render Sets
+
+<div class="release-feature-authors">authors: Rob Swain (@superdump), @james-j-obrien, @inodentry</div>
+
+The order of draws needs to be known for some methods of instanced draws so that the data can be laid out, and looked up in order. For example, when per-instance data is stored in an instance-rate vertex buffer.
+
+The render set order before 0.12 caused some problems with this as data had to be prepared before knowing the draw order. The previous order of sets was:
+
+![RenderSets-0.11](RenderSets-0.11.svg)
+
+This constraint was most visible in the sprite batching implementation that skipped `Prepare`, sorted and prepared data in `Queue`, and then after being sorted again alongside 2D meshes and other entities in the `Transparent2d` render phase, possibly had its batches split to enable drawing of those other entities.
+
+The ordering of the sets also created some confusion about when bind groups should be created. Bind groups were meant to be created in `Prepare`, but sometimes they had to be created in `Queue` to ensure that some preparation had completed.
+
+The new render set order in 0.12 is:
+
+![RenderSets-0.12](RenderSets-0.12.svg)
+
+`PrepareAssets` was introduced because we only want to queue entities for drawing if their assets have been prepared. Per-frame data preparation still happens in the `Prepare` set, specifically in its `PrepareResources` subset. That is now after `Queue` and `Sort`, so the order of draws is known. This also made a lot more sense for batching, as it is now known at the point of batching whether an entity that is of another type in the render phase needs to be drawn. Bind groups now have a clear subset where they should be created - `PrepareBindGroups`.
+
+### BatchedUniformBuffer and GpuArrayBuffer
+
+OK, so we need to put many pieces of data of the same type into buffers in a way that we can bind them as few times as possible and draw multiple instances from them. How can we do that?
+
+In 0.11 per-instance `MeshUniform` data is stored in a uniform buffer with each instance's data aligned to a dynamic offset. When drawing each mesh entity, we update the dynamic offset, which is close to rebinding. It looks like this:
```

Comment:
I don't understand what "which is close to rebinding" means

Reply:
A buffer binding is telling the graphics API that that buffer, at that memory address, and with that size needs to be accessible for the duration of execution of subsequent draws until the binding is changed. A dynamic offset is just a byte offset to the memory address of the start of the binding. In that sense, updating the dynamic offset, while a bit cheaper through some graphics APIs/drivers than updating the entire binding, is close to just updating that entire binding.

Reply:
I just clarified this with "close in cost"